qiskit/ignis/measurement/discriminator/iq_discrimination.py

```diff
@@ -0,0 +1,264 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+from typing import Union, List
+
+import numpy as np
+from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
+from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
+from sklearn.preprocessing import StandardScaler
+
+from qiskit.ignis.characterization.fitters import BaseFitter
+from qiskit.exceptions import QiskitError
+from qiskit.result.models import ExperimentResult
+from qiskit.result import postprocess, Result
+from qiskit.pulse.schedule import Schedule
+
+
+class ScikitIQDiscriminationFitter(BaseFitter):
+    """
+    IQDiscriminatorFitter takes IQ level 1 data produced by calibration
+    measurements with a known expected state. It fits a discriminator
+    that can be used to produce level 2 data, i.e. counts of quantum states.
+    """
+
+    def __init__(self, cal_results: Union[Result, List[Result]],
+                 qubits: List[int], expected_states: Union[List[str], str],
```

Comment:
Since this is really operating on input data from the `memory_slots` (which may be out of order) rather than input qubits isn't this something more akin to a `training_data_mask`. Ie., consider the case of multiple measurements being made on the same qubit. The input it seems would really need to be something like `{qubit_idx: [memoryslot_idx0, memoryslot_idx1, ...]}`. 

I apologize if one of your earlier proposals handled this case and I overlooked it.

Reply:
I think we talked about this on slack and mentioned a `GroupDiscriminator`. The current behaviour of `ScikitIQDiscriminationFitter` (and its children) is
- There is one discriminator (e.g. an LDA) in the class.
- This discriminator does the following:
    **a.** Select the IQ data out of the memory for the `self._qubits` qubits (i.e. a data mask really).
    **b.** Discriminate this data. This means assign a label, i.e. a `str`, to each IQ point.

That's it the discriminator does nothing more. There is no memory assignment to a `Result` instance.

The level 2 memory is assigned in the `filter` class. 

Maybe we should implement a `GroupIQDiscriminator` and a corresponding filter class that has as many `ScikitIQDiscriminationFitter`'s as needed and has a data structure that maps qubit indices to memory slots and says what `training_data_mask` should be used for which qubit?

One justification for this is that it keeps `ScikitIQDiscriminationFitter` conceptually simple: it has one discriminator that maps some IQ data to a label. The complexity of qubit indices and memslots and training data masks would then be in `GroupIQDiscriminator`.

Reply:
I think the issue I was trying to hint at, is that there is no guarantee that `memory_slot[0]` is where the result of qubit[0] is stored. We must either choose a convention or make this configurable. If we think of this as a fitting problem, one could provide a data mask for which memory slots to select, and then a target states (or qubits) for which states to discriminate.  

The grouped discriminator could still go on top of this (I think).

Reply:
OK. I think I see what you are getting at: `get_memory(...)` returns `np.ndarray[shots, memory_slots]` for level 1 data single shot and there is no guarantee that, e.g., `get_memory(...)[:, 0]` corresponds to the IQ data of qubit 0. Correct?

Is this behaviour intended? How do we know where to find the data for qubit X then?